test/hotspot/jtreg/compiler/c2/TestReduceAllocationAndNestedScalarized.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8331194
+ * @summary Check that Reduce Allocation Merges doesn't crash when an input
+ *          of the Phi is not the _current_ output of the Phi but said input
+ *          needs to be rematerialized because it's used regardless of the
+ *          Phi output.
+ * @requires vm.bits == 64 & vm.flagless & vm.compiler2.enabled
+ * @run main/othervm -XX:CompileCommand=dontinline,*TestReduceAllocationAndNestedScalarized*::test
+ *                   -XX:CompileCommand=compileonly,*TestReduceAllocationAndNestedScalarized*::test
+ *                   -XX:CompileCommand=compileonly,*Picture*::*init*
+ *                   -XX:CompileCommand=compileonly,*Point*::*init*
+ *                   -XX:CompileCommand=exclude,*Unloaded*::*
+ *                   -XX:-TieredCompilation
+ *                   -XX:-UseCompressedOops
+ *                   -Xcomp
+ *                   -server
+ *                   compiler.c2.TestReduceAllocationAndNestedScalarized
```

Comment:
Drive by comment: are all these `requires` absolutely necessary? Could there be a run without them? Is the test expected to fail with graal or C1? Or any external flags?

Could we have a run without some of the flags you list here? Just so that we have wider coverage, especially in combination with other flags that can be passed from the outside.

Reply:
This test is intended to "test" a transformation in C2, it won't fail if run in C1, but it won't actually test what's intended for. The `flagless` option is to make sure that C2 produce an IR graph that will trigger the optimization, same reasoning for all the "CompileCommand" options. I may remove the "vm.bits" part, I included it because of the "-UseCompressedOops".